File: cpp/samchon/protocol/service/Service.cpp
#include <samchon/protocol/service/Service.hpp>

#include <samchon/protocol/service/User.hpp>
#include <samchon/protocol/service/Client.hpp>

using namespace std;
using namespace samchon;
using namespace samchon::library;
using namespace samchon::protocol;
using namespace samchon::protocol::service;

Service::Service()
{
}
Service::~Service()
{
}

void Service::sendData(shared_ptr<Invoke> invoke)
{
	client->sendData(invoke);
}

auto Service::__keep_alive() const -> pair<shared_ptr<User>, shared_ptr<Client>>
{
	return client->__keep_alive();
}

File: ts/src/samchon/protocol/service/Service.ts
/// <reference path="../../API.ts" />

namespace samchon.protocol.service
{
	/**
	 * @author Jeongho Nam <http://samchon.org>
	 */
	export abstract class Service 
		implements protocol.IProtocol
	{
		/**
		 * @hidden
		 */
		private client_: Client;
		
		/**
		 * @hidden
		 */
		private path_: string;

		/* ------------------------------------------------------------------
			CONSTRUCTORS
		------------------------------------------------------------------ */
		/**
		 * Default Constructor.
		 */
		public constructor(client: Client, path: string)
		{
			this.client_ = client;
			this.path_ = path;
		}

		/**
		 * Default Destructor.
		 */
		public destructor(): void
		{
		}
		
		/* ------------------------------------------------------------------
			ACCESSORS
		------------------------------------------------------------------ */
		/**
		 * Get client.
		 */
		public getClient(): Client
		{
			return this.client_;
		}

		/**
		 * Get path.
		 */
		public getPath(): string
		{
			return this.path_;
		}

		/* ------------------------------------------------------------------
			MESSAGE CHAIN
		------------------------------------------------------------------ */
		public sendData(invoke: Invoke): void
		{
			return this.client_.sendData(invoke);
		}

		public abstract replyData(invoke: Invoke): void;
	}
}